### ISSUE TITLE
Incorrect boolean expression inferance

### ISSUE THREAD

**Issue by kong13661** [microsoft/pylance-release#2428]
## Environment data

-   Language Server version: 2020.2.4
-   OS and version: Ubuntu 20.04
-   Python version (& distribution if applicable, e.g. Anaconda): 3.9, Miniconda

## Expected behaviour

![image](https://user-images.githubusercontent.com/28509442/156383441-8915c46c-99a0-437f-add4-bb7d25b43c82.png)
This is apart code of class `Checkpoint`. Look at the code:

```
if not (callable(save_handler) or isinstance(save_handler, BaseSaveHandler)):
    raise TypeError("Argument `save_handler` should be callable or inherit from BaseSaveHandler")
```

If `callable(save_handler) or isinstance(save_handler, BaseSaveHandler)` is False, it will raise a error. But in fact, the type of argument `save_handler` is `callable` or `BaseSaveHandler`, and it will be True. 

![image](https://user-images.githubusercontent.com/28509442/156381279-3c69599e-ecd6-4cfb-8f46-3bec5f1c091b.png)
This is the code instantiating the class `Checkpoint`. If pylance works properly, it will infer the following code accessible.

## Actual behaviour

However, pylance seems to believe the expression to be False.

![image](https://user-images.githubusercontent.com/28509442/156381796-79c0fd10-fd4a-4b3c-958e-53e512ec18c1.png)

Pylance believes the following code is not accesible.

## Logs

```
XXX
```

## Code Snippet / Additional information

In fact, if I rewrite the code with an equivalent expression:

![image](https://user-images.githubusercontent.com/28509442/156382252-74be6373-c94a-4c16-a8b2-0cd6261792be.png)

it will work as expected.

![image](https://user-images.githubusercontent.com/28509442/156383178-73b31f65-0cd0-4372-92dc-a38f2b2bfbdd.png)

This bug is induced in version 2022.1.0

**Comment by erictraut** [comment]
Thanks for reporting. Could you produce a minimal self-contained sample that reproduces the problem? Partial screen shots of your code doesn't provide enough information. If the minimal sample depends on symbols imported from a library, please show those imports in the code snippet and provide the versions of the libraries that you're importing. Thanks!

**Comment by kong13661** [comment]
> Thanks for reporting. Could you produce a minimal self-contained sample that reproduces the problem? Partial screen shots of your code doesn't provide enough information. If the minimal sample depends on symbols imported from a library, please show those imports in the code snippet and provide the versions of the libraries that you're importing. Thanks!

```python
from typing import Union, Callable


class A:
    def __call__(self) -> None:
        pass


class B:

    def __init__(
        self,
        save_handler: Union[Callable, A],
    ):

        if not (callable(save_handler) or isinstance(save_handler, A)):
            raise Exception


c = B(10)
d = 10
```

This is the minimal code found. The class `A` must contain the member function `__call__`, and `callable` is needed in the boolean expression `if not (callable(save_handler) or isinstance(save_handler, A)):`. I don't know if there is other way to reproduce this bug.

![image](https://user-images.githubusercontent.com/28509442/156509541-81bd5a9d-d532-48e1-9f35-5afd7f9d03d5.png)

Replace ` if not (callable(save_handler) or isinstance(save_handler, A)):` by an equivalent expression:

```python
if not callable(save_handler) and not isinstance(save_handler, A):
```

![image](https://user-images.githubusercontent.com/28509442/156509495-9e7910bc-974d-4ad7-bc73-b7a9a98aaa8b.png)

It works properly.

Need I rewrite my issue above?

**Comment by erictraut** [comment]
Thanks for the repro. This bug will be fixed in the next release.

**Comment by kong13661** [comment]
> Thanks for the repro. This bug will be fixed in the next release.

Thanks a lot.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.3.1, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202231-9-March-2022)
